Source/Plugins/Core/com.equella.core/scalasrc/com/tle/core/cloudproviders/CloudProviderRegistrationService.scala

```diff
@@ -0,0 +1,204 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.tle.core.cloudproviders
+
+import cats.data.ValidatedNec
+import cats.implicits._
+import com.tle.beans.newentity.Entity
+import com.tle.common.i18n.CurrentLocale
+import com.tle.core.guice.Bind
+import com.tle.core.newentity.service.EntityService
+import com.tle.core.replicatedcache.ReplicatedCacheService
+import com.tle.core.validation.{EntityStdEdits, EntityValidation}
+import com.tle.legacy.LegacyGuice
+import com.tle.web.DebugSettings
+import io.circe.parser._
+import io.circe.syntax._
+import org.slf4j.LoggerFactory
+import java.time.Instant
+import java.util.UUID
+import java.util.concurrent.TimeUnit
+import javax.inject.{Inject, Singleton}
+import scala.collection.JavaConverters._
+
+@Bind
+@Singleton
+class CloudProviderRegistrationService {
+  @Inject var entityService: EntityService = _
+
+  def tokenCache: ReplicatedCacheService.ReplicatedCache[String] =
+    LegacyGuice.replicatedCacheService.getCache[String]("cloudRegTokens", 100, 1, TimeUnit.HOURS)
+  val FieldVendorId    = "vendorId"
+  val RefreshServiceId = "refresh"
+  val typeId           = "cloudprovider"
+  val Logger           = LoggerFactory.getLogger(getClass)
+
+  // Transform an Entity's custom data from String to CloudProviderData.
+  private def extractData(data: String): ValidatedNec[EntityValidation, CloudProviderData] =
+    decode[CloudProviderData](data)
+      .leftMap(err => EntityValidation("data", err.getMessage))
+      .toValidatedNec
+
+  // Build a CloudProviderInstance by Entity and CloudProviderData
+  private def toInstance(entity: Entity, data: CloudProviderData): CloudProviderInstance = {
+    CloudProviderInstance(
+      id = UUID.fromString(entity.id.uuid),
+      name = entity.name,
+      description = Option(entity.description),
+      vendorId = data.vendorId,
+      baseUrl = data.baseUrl,
+      iconUrl = data.iconUrl,
+      providerAuth = data.providerAuth,
+      oeqAuth = data.oeqAuth,
+      serviceUrls = data.serviceUrls,
+      viewers = data.viewers
+    )
+  }
+
+  // Save an Entity with provided values, and return the CloudProviderInstance representing
+  // this Entity.
+  private def save(entity: Entity,
+                   values: (EntityStdEdits, CloudProviderData)): CloudProviderInstance = {
+    val (edits, data) = values
+    entity.data = data.asJson.noSpaces
+    entity.modified = Instant.now
+    entity.name = edits.name
+    entity.nameStrings = edits.nameStrings.asJson.noSpaces
+    entity.description = edits.description.getOrElse("")
+    entity.descriptionStrings = edits.descriptionStrings.map(_.asJson.noSpaces).orNull
+    entityService.createOrUpdate(entity)
```

Comment:
Hmmm, this got lost in my first reading - indeed, I wondered why this was called `save` as I thought it didn't persist anything.

Maybe to make it jump out:

```suggestion

    // Persist the new Entity
    entityService.createOrUpdate(entity)
```

Reply:
Ah yes it's to persist an entity. I will add this comment.